File: backend/app/schemas/impact_timeline.py
from datetime import date

from pydantic import BaseModel


class ImpactSnapshot(BaseModel):
    recorded_at: date
    carbon_captured_tco2e: float
    biodiversity_score: float
    vegetation_cover_percent: float
    species_count: int


class ImpactChanges(BaseModel):
    carbon_pct: float
    biodiversity_points: float
    vegetation_pct: float
    species_count_delta: int
    nature_health_score_delta: float | None


class ImpactTimelineRead(BaseModel):
    """Restoration Impact Timeline -- derived entirely from existing
    `site_analytics` rows (no new table). `summary` is template-filled from
    `changes`; it is NOT written by an LLM.
    """

    first: ImpactSnapshot
    latest: ImpactSnapshot
    changes: ImpactChanges
    summary: str

File: frontend/src/types/impactTimeline.ts
/** One point-in-time snapshot of the four headline metrics used by the impact timeline. */
export interface ImpactTimelineSnapshot {
  recorded_at: string
  carbon_captured_tco2e: number
  biodiversity_score: number
  vegetation_cover_percent: number
  species_count: number
}

export interface ImpactTimelineChanges {
  carbon_pct: number
  biodiversity_points: number
  vegetation_pct: number
  species_count_delta: number
  nature_health_score_delta: number | null
}

export interface ImpactTimelineResponse {
  first: ImpactTimelineSnapshot
  latest: ImpactTimelineSnapshot
  changes: ImpactTimelineChanges
  /**
   * Template-filled from `changes` by the backend — explicitly NOT written by an LLM, see
   * docs/FEATURE_CONTRACT_V2.md "Restoration impact timeline".
   */
  summary: string
}
